homeassistant/components/repetier/sensor.py

```diff
@@ -0,0 +1,294 @@
+                                        printers[id]['slug'] +
+                                        '_current_job')
+                if not state:
+                    _LOGGER.debug("Job percentage sensor initiating...")
+                    new_sensor = RepetierSensor('percentage',
+                                                {'name': printers[id]['slug'],
+                                                 'id': id,
+                                                 'server': {'url': url,
+                                                            'port': port,
+                                                            'apikey': apikey},
+                                                 'sens_type': sens_type})
+                    devices.append(new_sensor)
+            elif sens_type == "Time Remaining":
+                state = hass.states.get('sensor.' +
+                                        printers[id]['slug'] +
+                                        '_current_job_remaining')
+                if not state:
+                    _LOGGER.debug("Time remaining sensor initiating...")
+                    new_sensor = RepetierSensor('remaining',
+                                                {'name': printers[id]['slug'],
+                                                 'id': id,
+                                                 'server': {'url': url,
+                                                            'port': port,
+                                                            'apikey': apikey},
+                                                 'sens_type': sens_type})
+                    devices.append(new_sensor)
+            elif sens_type == "Time Elapsed":
+                state = hass.states.get('sensor.' +
+                                        printers[id]['slug'] +
+                                        '_current_job_elapsed')
+                if not state:
+                    _LOGGER.debug("Time elapsed sensor initiating...")
+                    new_sensor = RepetierSensor('elapsed',
+                                                {'name': printers[id]['slug'],
+                                                 'id': id,
+                                                 'server': {'url': url,
+                                                            'port': port,
+                                                            'apikey': apikey},
+                                                 'sens_type': sens_type})
+                    devices.append(new_sensor)
+
+        add_entities(devices, True)
+
+    _addDevice()
+
+
+class RepetierSensor(Entity):
+    """Class to create and populate a Repetier Sensor."""
+
+    def __init__(self, type, data):
+        """Init new sensor."""
+        self._name = data['name']
+        self._sens_type = data['sens_type']
+        self._type = type
+        self._server = data['server']
+        self._id = data['id']
+
+        if type == 'nozzle':
```

Comment:
But even when splitted, you need to check on type using if elif, Python doesn't support switch case.

Reply:
If we simplify the options by splitting into multiple classes, remaining options can go into a dictionary that we look up.

Reply:
I'm lost - I have no idea what you have in mind.

Reply:
I suggest we fetch the new data in a central api instance in the component and store the new data in a state cache, which is a dict, on the api instance. We make sure to model the state cache to store each sensor_type data dict under the sensor_type key and use another `data_key` specific for the instance of a sensor_type to mark the sensor instance data.

It looks like there could be multiple instances of each sensor_type, at least for nozzles and beds. With the sensor_type and the data_key we should be able to get the data specific for a sensor instance from the state cache.

Then we just have to pass the `sensor_type` and `data_key` in the discovery_info, to allow the sensor entities to get the correct data when we signal them to update.

Something like this, taking the bed sensor as example:

```py
REPETIER_API = 'repetier_api'
UPDATE_SIGNAL = 'repetier_update_signal'


def setup_platform(hass, config, add_entities, discovery_info=None):
    """Set up the available Repetier Server sensors."""
    if discovery_info is None:
        return

    api = hass.data[REPETIER_API]

    data_key = discovery_info['data_key']
    name = discovery_info['name']
    printer_id = discovery_info['printer_id']
    sensor_type = discovery_info['sensor_type']

    sensor_map = {
        'bed_temperature': RepetierBedSensor,
        'nozzle_temperature': RepetierNozzleSensor,
        'current_state': RepetierStateSensor,
        'job_percentage': RepetierJobSensor,
        'time_remaining': RepetierRemainingSensor,
        'time_elapsed': RepetierElapsedSensor,
    }

    sensor_class = sensor_map[sensor_type]
    entity = sensor_class(api, data_key, name, printer_id, sensor_type)
    add_entities([entity], True)


class RepetierSensor(Entity):
    """Class to create and populate a Repetier Sensor."""

    def __init__(self, api, data_key, name, printer_id, sensor_type):
        """Init new sensor."""
        self._api = api
        self._available = False
        self._name = name
        self._sensor_type = sensor_type
        self._data_key = data_key
        self._id = printer_id
        self._state = None
        self._attributes = {}

    @property
    def available(self) -> bool:
        """Return True if entity is available."""
        return self._available

    @property
    def device_state_attributes(self):
        """Return sensor attributes."""
        return self._attributes

    @property
    def name(self):
        """Return the name of the sensor."""
        return self._name

    @property
    def unit_of_measurement(self):
        """Return the unit of measurement of this entity, if any."""
        return SENSOR_TYPES[self._sensor_type][1]

    @property
    def icon(self):
        """Icon to use in the frontend."""
        return SENSOR_TYPES[self._sensor_type][2]

    @callback
    def update_callback(self):
        """Get new data and update state."""
        self.async_update_ha_state(True)

    async def async_added_to_hass(self):
        """Connect update callbacks."""
        async_dispatcher_connect(
            self.hass, UPDATE_SIGNAL, self.update_callback)

    def _get_data(self):
        """Return new data from the api cache."""
        data = self._api.data.get(self.sensor_type, self._data_key)
        if data is None:
            _LOGGER.debug(
                "Data not found for %s and %s",
                self.sensor_type, self._data_key)
            self._available = False
            return None
        self._available = True
        return data


class RepetierBedSensor(RepetierSensor):
    """Class to create and populate a Repetier Bed Sensor."""

    @property
    def state(self):
        """Return sensor state."""
        if self._state is None:
            return None
        return round(self._state, 2)

    def update(self):
        """Update the sensor."""
        data = self._get_data()
        if data is None:
            return
        temp_set = data['tempset']
        temp = data['temp']
        _LOGGER.debug(
            "Bed %s Setpoint: %s, Temp: %s", self._data_key, temp_set, temp)
        self._attributes['setpoint'] = temp_set
        self._state = temp


class RepetierNozzleSensor(RepetierSensor):
    """Class to create and populate a Repetier Nozzle Sensor."""


class RepetierStateSensor(RepetierSensor):
    """Class to create and populate a Repetier State Sensor."""


class RepetierJobSensor(RepetierSensor):
    """Class to create and populate a Repetier Job Sensor."""


class RepetierRemainingSensor(RepetierSensor):
    """Class to create and populate a Repetier Time Remaining Sensor."""


class RepetierElapsedSensor(RepetierSensor):
    """Class to create and populate a Repetier Time Elapsed Sensor."""
```

Reply:
I think I follow you here, but....

REPETIER_API = 'repetier_api'
That would be the instance from the PyPi module?!

UPDATE_SIGNAL = 'repetier_update_signal'
What is this?

api = hass.data[REPETIER_API]
How do hass.data get to contain REPETIER_API? I don't see where you set that?!
Do you associate this in __init__.py?

I guess I'll just have to try a 3rd iteration of the plugin :/ Perhaps I'll get lucky one day.

Reply:
The api is class that we write in the component that handles the api interaction via the library client.

So in `setup` in the component we instantiate the api class and save it in hass.data. I didn't write this to in the example.

The api class should handle all api interactions and transform the incoming data into the correct shape and store it in the state cache dict.

Reply:
Do we have any documentation available for these functions?

I have no idea what these are and what they are expected to do:

self.async_schedule_update_ha_state(True)
async_dispatcher_connect(self.hass, UPDATE_SIGNAL, self.update_callback)

I'll try to make a new version of this component but can't promise it'll be with these callbacks. I can find no documentation of this anywhere - hence it's pretty hard to make use of this.

Reply:
And I have been looking at the MySensors component - that only made me more confused.

Reply:
https://developers.home-assistant.io/docs/en/entity_index.html

Reply:
The dispatcher connect function connects a target function with a signal. The dispatcher send function sends the signal to call the target function passing possible arguments to the tarfet function.

Reply:
`async_schedule_update_ha_state`, and the sync version `schedule_update_ha_state`, is the entity method to schedule a state update. Passing `True` will make sure `update` or `async_update` is called before the state update.

Reply:
Currently trying to get it up and running like this.

`sensor_map` would that be the same values as `SENSOR_TYPES` from `__init__.py`?

Reply:
Yes, it should use the keys of that constant dictionary. But extend it so that the different temperature sensors, nozzle and bed, have different sensor types, as in my example. Flat structures are generally better than nested structures.

Reply:
Okay - so far so good.
All sensors are working and returning correct data.

Only the callback part adding new sensors as they become available are not made yet - have yet to find what I need more :)